### PR TITLE
feat: rewrite js rpg project

### PR DESCRIPTION
**Pull request by naomi-lgbt**
Checklist:

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

Big enough to crash GitHub probably...

Rewrites the JS RPG project to:
- Have the camper write the HTML and CSS (for retention and to demonstrate how the JS interacts with this)
- Leverage the updated multi-file editor (instead of everything in one file with script/style tags)
- Provide clarity in some of the step instructions
- More granular tests

### CODE REVIEW

**Comment by Sboonny** [comment]
I have committed a test example for step 1, I am wondering if this acceptable and may I do more, or it's better for me to wait for it to be ready for review? @nhcarrigan

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md:10]:
```suggestion
Similar to your `#stats` element, your `#monsterStats` element needs two `span` elements. Give them the class `stat` and give the first element the text `Monster Name: ` and the second the text `Health: `. After the text in each, add a `strong` element with an empty nested `span` element.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a24190868ca51c0b6e83c7.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a24190868ca51c0b6e83c7.md:10]:
```suggestion
Hide your `#monsterStats` element for now but give it the same `border` and `padding` as your `#stats` element. Set its color to `white` and give it a `red` background.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a39f5a5790eb27c1e5d4bf.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a39f5a5790eb27c1e5d4bf.md:12]:
```suggestion
Add your `console.log("Hello World");` line to this file, and see it appear in your console.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3a0a3c0a4b32915d26a6e.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3a0a3c0a4b32915d26a6e.md:12]:
```suggestion
In JavaScript, a <dfn>variable</dfn> is used to hold a value. To use a variable, you must first <dfn>declare</dfn> it. For example, to declare a variable called `camperbot`, you would write:
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3a488b24fb32b91155d56.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3a488b24fb32b91155d56.md:10]:
```suggestion
Variables can be <dfn>assigned</dfn> a value. When you do this while you declare it, this is called <dfn>initialization</dfn>. For example:
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3b365f1cdeb33efc2502e.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3b365f1cdeb33efc2502e.md:10]:
```suggestion
You have been declaring your variables with the `var` keyword. However, in modern JavaScript it is best practice to use the `let` keyword instead. This fixes several unusual behaviors with `var` that can make your code difficult to debug.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3b79d520a7f3d0e25afd6.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3b79d520a7f3d0e25afd6.md:18]:
```suggestion
Note that `h1` is a string and matches the CSS selector you would use. Create a `button1` variable and use `querySelector()` to assign it your element with the `id` of `button1`.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8a929e4260d08093756d2.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8a929e4260d08093756d2.md:10]:
```suggestion
Create a variable called `locations` and set it to an empty array.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8ab0e27cbaf0b54ba8a42.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8ab0e27cbaf0b54ba8a42.md:10]:
```suggestion
Just like array values, object properties are separated by a comma. Add a comma after your `name` property and add a `button text` property with the value of an empty array. Note that because the property name has more than one word, you'll need to surround it in quotes.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b0b5053f16111b0b6b5f.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b0b5053f16111b0b6b5f.md:10]:
```suggestion
Now you can consolidate some of your code. Start by copying the code from inside the `goTown` function and paste it into your `update` function. Then, remove all the code from inside the `goTown` and `goStore` functions.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b1762b7775124622e1a3.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b1762b7775124622e1a3.md:10]:
```suggestion
Instead of assigning the `innerText` and `onclick` properties to specific strings and functions, the `update` function will use data from the `location` that is passed into it. First, that data needs to be passed. Inside the `goTown` function, call the `update` function. Here is an example of calling a function named `myFunction`: `myFunction();`.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b6536156c51500739b41.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b6536156c51500739b41.md:12]:
```suggestion
This is called <dfn>bracket notation</dfn>. Values in an array are accessed by index. Indices are numerical values and start at 0 - this is called zero-based indexing. `arg[0]` would be the first element in the `arg` array.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b711ab7a12161c7d9b67.md
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a8b711ab7a12161c7d9b67.md:10]:
```suggestion
Now your `update` function needs to use the argument you pass into it. Inside the `update` function, change the `button1.innerText` assignment to be `location["button text"]`. That uses bracket notation to get the `button text` property of the `location` object passed into the function.
```

**Review by Sembauke — DISMISSED** [review]

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a7bf807a92321df71ee57d.md
Comment (nhcarrigan) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a7bf807a92321df71ee57d.md:10]: Yes 🙃

**Comment by nhcarrigan** [comment]
@Sboonny thanks for your interest in helping!

In general, draft PRs are used to solicit early feedback before a feature/change is actually ready to merge. As a practice, we generally don't commit to others' PRs unless there's been prior discussion and a decision to collaborate. (:

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a7c23e6b511f22ed71197a.md
Comment (jdwilkin4) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a7c23e6b511f22ed71197a.md:10]: We should add a code example here because beginners won't know intuitively where to place the backslashes.

**Review by jdwilkin4 — CHANGES_REQUESTED** [review]
Here is some early feedback from me 👍

**Review by Sboonny — COMMENTED** [review]
This is one of the bugs I added, sorry about that. The seed code has `<html lang="en">` but it's testing for `<html>`

**Review by sidemt — COMMENTED** [review]
I took a look at first 10 steps.

**Review by jdwilkin4 — COMMENTED** [review]
Left some comments for the first 15 steps

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3cdb11478a34ff4a6470d.md
Comment (jdwilkin4) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3cdb11478a34ff4a6470d.md:12]: In the console I am getting this error, because the script is loaded before our button elements because we placed the script element in the head. 
Screen Shot 2022-07-10 at 8 17 29 PM

**Review by jdwilkin4 — COMMENTED** [review]
These are my comments for steps 15 - 111.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a1166ed9a56d439c0770e7.md
Comment (Sboonny) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a1166ed9a56d439c0770e7.md:30]:
```suggestion
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a2509ba163e020bb9d84ea.md
Comment (Sboonny) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a2509ba163e020bb9d84ea.md:17]:
```suggestion
```

**Review by Sboonny — COMMENTED** [review]
Console.log shall not pass 🧙‍♂️

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62aa2aec2f09d454253aad6c.md
Comment (gikf) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62aa2aec2f09d454253aad6c.md:10]:
```suggestion
Since you subtracted health from the player, you need to check `if` the player's `health` is less than or equal to `0`. If it is, call the `lose` function.
```

**Review by gikf — COMMENTED** [review]
It's pretty fun.

Couple of nitpicks in the specific suggestions.

There might be slight issue with the secret game. It can be played multiple times on a single _game encounter_, what makes it a bit game breaking with winning 20 gold for a win, with a rather small risk or losing 10 health.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md
Comment (Sboonny) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md:40]: I don't know what is happening, but you are getting `Health:` here instead of `Health: `

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a24128d5e8af1b47ad1aab.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a24128d5e8af1b47ad1aab.md:17]: I do not want to have too many people reviewing one PR. But:

Would it help if the helper function had a flag to automatically search for permutations of selectors like this?

**Comment by moT01** [comment]
I see a failing test here - didn't look into it. Changing the label.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md
Comment (naomi-lgbt) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md]: Changing this to `textContent` to avoid automatically trimming the blank space seems to fix it?

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a24128d5e8af1b47ad1aab.md
Comment (naomi-lgbt) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a24128d5e8af1b47ad1aab.md:17]: > I do not want to have too many people reviewing one PR. But:

This is the first JS project so it's kind of important - no such thing as too many reviewers.

Yes, that might be a helpful feature, but low priority. We are only having campers write the HTML/CSS for the first couple of projects (to link their previous learning to these new concepts) before we just start providing it all for them and focus entirely on the JS.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md
Comment (Sboonny) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a23d1c5f1c93161f3582ae.md]: https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent, It seems it will get `<strong>` and `<span>` too, but it will work. 

I think it's better to change the test to check for `Health:` because it's JavaScript steps, so we can go lenient.

We can go regex too, but I prefer nothing over the regex.

**Review by moT01 — COMMENTED** [review]
Okay, wow - this is a beast of a project. Overall, it's pretty awesome @naomi-lgbt 🎉 I won:

Screen Shot 2022-08-15 at 2 56 35 PM

 I went through it all, here's everything I wrote down - apologies if some of it is a little scattered...

- Step 1: Doesn't say to add `html` element.

- Step 12: Could maybe be broken up into two steps - step 1: add the style -> step 2: hide it. I hid it first, so when I added the styles, I didn't see anything. Also, I used `visibility: hidden` and the test failed. The instructions implied to me to use that, but they expect `display: none;` could maybe allow either of those to pass?

- Step 15: Maybe don't need to mention the dev tools and just have them use our console to start? Instead of "You should see the text in the console" maybe instruct them to open it.
Could mention that the script runs and gets printed to the console and that when you run the tests, you will see a second log from those.

- Step 19: Has regex `/var\sxp\s?=\s?0;/` so this didn’t pass: `var xp =0 ;` I would try to use the "zero or more" and "one or more" flags. e.g. `/var\s+xp\s*=\s*0\s*;/` - more notes about this at the bottom.

- Step 25: Likely similar to step 19 - I didn’t check the tests, but it mentions you can use single/double quotes or backticks - only double quotes pass. 

- Step 28: Add `let h1 =` to example

- Step 29: Could maybe add a few steps where they console.log() the element - see it’s not there, then have them move the script tag.

- Step 31: Adds a lot of new variables - could maybe have another step before this where they just add `button2` and `button3` - then add a whole bunch after they have the hang of it a little more. Could also maybe split the last variable, `monsterHealthText`, into its own step and explain that you can only have one variable name in the scope.

- Step 39: Could maybe use a better example, maybe `button1.innerText = …`

- Step 63: Seed code - there's two `"button functions"` properties in the `locations` objects. First one should be `"button text"`. There may be other steps with this issue - I didn't look very hard, but I didn't notice any.

- Step 67: Seed code automatically adds the comma before adding another object to `locations[]` - I would have campers add that.

- Step 72: I struggled with this one for a minute - I put gold and health as strings and not the variables 🤦

- Step 87: Could maybe use a slight reword - it doesn't really say that I need a period at the end of the sentence. More notes about this at the bottom.

- Step 103: Similar to step 87 - It doesn't say I need a space at the beginning - also not shown in the code blocks

- Step 132: Little confused - I am supposed to use the sixth object in `locations`, but we only created 5 - but the seed code has a sixth - test passed with [4] instead of [5] - looks like we end up adding the sixth object later.

- Step 138: Maybe typo: defeat -> defeated?

- Step 144: Typo - retrun -> return

Here are some more general things I noticed throughout the project…

Mentioned on steps 19 and 25 above. I'm not sure how many regex tests there are throughout, certainly more than these two steps - but I noticed that they were quite strict with regards to spacing and which quotes you can use to enclose a string. It seemed to improve as I got further into the project.

There were a number of code blocks at the end of a sentence. I don't mind that alone, but many of them end with a period - then have another period to end the instruction/hint - e.g. "Your second `locations` object should have a text property with the value of `You enter the store.`." I feel like we should just leave out that last period in these cases.

I was back and fourth on this a little - I worry that the project might be too difficult. Or at least, difficult in areas. There’s spots early on where it seems too advanced for beginners and not granular enough - and spots later on that are more granular and should work well. Not sure what to do about it, just mentioning it.

Many of the variable names are "text" - e.g. `healthText`. I commonly use "el" or "element" - e.g `healthElement`. I think that might help campers understand better.

I would have liked to see more steps where we have campers interact with the preview, or look at the console to see more what is happening. Maybe "click the button now to see the function run" or something. Although, there was some of that in there, so I can't complain too much.

Mentioned in steps 87 and 103 above. I had a lot of issues around spacing when concatenating sentences. There wasn't much in the instructions or hints where I needed them.

Some of the language felt a little rough - e.g. "lose" and "kill". One steps says to add this: `You die. ☠️` 

<hr>

Many of the above things may be more of a preference - add/use them at your discretion @naomi-lgbt. I wouldn't say any of them should stop this from getting in, so let us know if you're ready.

I had a lot of fun going through it 🎉

**Comment by naomi-lgbt** [comment]
@moT01 thanks for the solid review - especially with how GitHub handles large PRs.

I'll circle back and hit this tomorrow!

**Comment by naomi-lgbt** [comment]
Okay, I should have covered most of your points now.

> I worry that the project might be too difficult.

I 100% agree. I've been pondering this for a bit, but honestly I feel like teaching DOM manipulation at the same time as JavaScript fundamentals is going to be inherently overwhelming. The problem is I'm not able to see a better approach while still taking the project-based structure.

**Comment by jdwilkin4** [comment]
>I feel like teaching DOM manipulation at the same time as JavaScript fundamentals is going to be inherently overwhelming.

We could try inserting a simpler smaller project before this rpg game and have it focus on the fundamentals with a little bit of DOM manipulation. 
Then we could place the rpg game second, and it could be a nice review of the fundamentals.

**Comment by moT01** [comment]
Is this ready for another look @naomi-lgbt?

**Comment by naomi-lgbt** [comment]
Yep!

**Review by Dario-DC — COMMENTED** [review]
I went through the project (I really enjoyed it) and I found several things that could be confusing as a JS beginner:
- It would be nice if the console could stay open. If you have a syntax error and you press the check your code button with the console closed, you don't realize what is going on since tests don't run. 
- I think the equality/inequality operator are not introduced (maybe I missed them) and the step 138 requires the use of strict equality
- step 29, in the code example `const` appears out of the blue (explained in next steps)
- step 74, there is a typo: `compount`
- step 115, the code below passes the text:
```
function goFight() {
  update(locations[3]);
  monsterHealth = monsters[fighting][health];
}
``` 
but `[health]` results in an undefined value.
- step 147, the syntax of the ternary operator with the return statement should be explained, my first approach was `... ? return... : return...`
- step 170, it could be helpful add an example of indexOf() syntax

**Comment by moT01** [comment]
Have your concerns been resolved here @Sembauke @jdwilkin4 ?

**Comment by jdwilkin4** [comment]
@moT01 

Yes, I have resolved all of my earlier comments. 👍

**Comment by naomi-lgbt** [comment]
> It would be nice if the console could stay open. If you have a syntax error and you press the check your code button with the console closed, you don't realize what is going on since tests don't run.

This is good feedback, but would be outside the scope of this PR.

**Comment by moT01** [comment]
Why is this test failing here, but not on any other PR's

**Comment by Sboonny** [comment]
I remember the same test failing in a typo pull request, I thought it was a fluke, but this is the second time I see it happening.

**Comment by moT01** [comment]
I updated the branch - maybe that will do something. I see no reason that test should be failing here.

**Comment by moT01** [comment]
Ahh, I was under the impression that it was the duplicate project titles that were making this fail, but I don't think that's it. It's an out of memory issue.

**Comment by moT01** [comment]
Looks like that more ram PR finally got the tests to pass. I wonder how close this puts us to the limit of that ram. We added one "more ram" PR and it was failing consistently (the upcoming changes test). Then we added that second "more ram" PR, and it passed once. As we add more of these projects, I'm guessing we will come across this issue again - I'm not sure though. We may need to rethink how we test the curriculum. Perhaps a stress test PR where we add another 500? 1000? challenges or something to see if we can find that limit before it becomes a problem.

I'm rerunning the tests once more to make sure it wasn't a fluke.

**Review by moT01 — APPROVED** [review]
Well, this has had quite a few reviews and adjustments - I fired it up once more and did a few steps, didn't go through it all again. I would be fine getting it in. LGTM 🎉

**Review by jdwilkin4 — APPROVED** [review]
I went through quite a few of the steps and I think this PR is ready to be merged.
LGTM 👍

**Comment by moT01** [comment]
Well, if you're satisfied @naomi-lgbt - I think we can get this in.

**Pull request merged**
